lib/std/padded_array.zig

```diff
@@ -0,0 +1,123 @@
+/// The array contains a maximum of `length` elements of type `T`. The length is determined by
+/// the first appearance of the `sentinel` value.
+///
+/// This variant can define the structure layout, so it can be used with external or packed variants.
+pub fn PaddedArrayExtra(comptime T: type, comptime length: usize, comptime sentinel: T, comptime layout: std.builtin.Type.ContainerLayout) type {
+    return switch (layout) {
+        .Auto => struct {
+            pub usingnamespace PaddedArrayImplementation(@This(), T, length, sentinel);
+            items: [length]T = [1]T{sentinel} ** length,
+        },
+        .Packed => packed struct {
+            pub usingnamespace PaddedArrayImplementation(@This(), T, length, sentinel);
+            items: [length]T = [1]T{sentinel} ** length,
+        },
+        .Extern => extern struct {
+            pub usingnamespace PaddedArrayImplementation(@This(), T, length, sentinel);
+            items: [length]T = [1]T{sentinel} ** length,
+        },
+    };
+}
+
+fn PaddedArrayImplementation(comptime Self: type, comptime T: type, comptime length: usize, comptime sentinel: T) type {
+    return struct {
+        fn Span(comptime Inner: type) type {
+            return switch (Inner) {
+                *const [length]T => []const T,
+                *[length]T => []T,
+                else => unreachable,
+            };
+        }
+
+        /// Returns the maximum number of items in this array.
+        pub fn capacity(self: Self) usize {
+            _ = self;
+            return length;
+        }
+
+        /// Returns the number of items in the array before the sentinel appears.
+        /// If no sentinel is present, the length of the array is returned.
+        pub fn len(self: Self) usize {
+            return std.mem.indexOfScalar(T, &self.items, sentinel) orelse length;
+        }
+
+        /// View the internal array as a slice whose size was previously set.
+        pub fn slice(self: anytype) Span(@TypeOf(&self.items)) {
+            return self.items[0..self.len()];
+        }
+
+        /// View the internal array as a constant slice whose size was previously set.
+        pub fn constSlice(self: *const Self) []const T {
+            return self.slice();
+        }
+
+        /// Copy the content of an existing slice.
+        pub fn fromSlice(m: []const T) error{Overflow}!Self {
+            if (m.len > length)
+                return error.Overflow;
+            var list: Self = .{};
+            std.mem.copy(T, &list.items, m);
+            return list;
+        }
+
+        /// Return the element at index `i` of the slice.
+        pub fn get(self: Self, i: usize) T {
+            return self.constSlice()[i];
+        }
+
+        /// Set the value of the element at index `i` of the slice.
+        pub fn set(self: *Self, i: usize, item: T) void {
+            self.slice()[i] = item;
+        }
```

Comment:
I suggest to remove these, thus encouraging the API user doing `x.slice()[i]` directly.